subprojects/architecture-test/src/test/java/org/gradle/architecture/test/KotlinCompatibilityTest.java

```diff
@@ -0,0 +1,216 @@
+            if (propertyAccessorType != null && (KotlinCompatibilityTest.isGetter(method, propertyAccessorType) || KotlinCompatibilityTest.isSetter(method, propertyAccessorType))) {
+                return new Accessor(propertyAccessorType, method);
+            }
+            return null;
+        }
+
+        private Accessor(PropertyAccessorType accessorType, JavaMethod method) {
+            this.accessorType = accessorType;
+            this.method = method;
+            this.isGetter = accessorType == PropertyAccessorType.IS_GETTER || accessorType == PropertyAccessorType.GET_GETTER;
+        }
+
+        public String getPropertyName() {
+            return accessorType.propertyNameFor(method.getName());
+        }
+
+        public JavaMethod getMethod() {
+            return method;
+        }
+
+        public boolean isGetter() {
+            return isGetter;
+        }
+
+        public boolean isSetter() {
+            return !isGetter;
+        }
+    }
+
+    private static class Accessors {
+        private final JavaClass owningClass;
+        private final String propertyName;
+        private final Set<JavaMethod> getters;
+        private final Set<JavaMethod> setters;
+
+        @Nullable
+        public static Accessors from(String propertyName, List<Accessor> accessors) {
+            Map<Boolean, List<Accessor>> gettersAndSetters = accessors.stream().collect(partitioningBy(Accessor::isGetter));
+            JavaClass owningClass = accessors.iterator().next().getMethod().getOwner();
+            Set<JavaMethod> getters = gettersAndSetters.get(Boolean.TRUE).stream().map(Accessor::getMethod).collect(toSet());
+            Set<JavaMethod> setters = gettersAndSetters.get(Boolean.FALSE).stream().map(Accessor::getMethod).collect(toSet());
+            if (!getters.isEmpty() && !setters.isEmpty()) {
+                return new Accessors(owningClass, propertyName, getters, setters);
+            }
+            List<Accessor> superclassAccessors = findAccessors(owningClass.getAllMethods()).get(propertyName);
+            if (superclassAccessors == null) {
+                return null;
+            }
+            if (!setters.isEmpty()) {
+                Set<JavaMethod> superclassGetters = superclassAccessors.stream().filter(Accessor::isGetter).map(Accessor::getMethod).collect(toSet());
+                if (!superclassGetters.isEmpty()) {
+                    return new Accessors(owningClass, propertyName, superclassGetters, setters);
+                }
+            } else {
+                Set<JavaMethod> superclassSetters = superclassAccessors.stream().filter(Accessor::isSetter).map(Accessor::getMethod).collect(toSet());
+                if (!superclassSetters.isEmpty()) {
+                    return new Accessors(owningClass, propertyName, getters, superclassSetters);
+                }
+            }
+            return null;
+        }
```

Comment:
:o: I would split this method into several explaining methods to make it more clear what happens